Add was_migration_trial flag for wpcom sites

zaerl opened the pull request:
This PR adds the `was_migration_trial` flag to the site data we return from the site REST APIs. This value will always be false for normal Jetpack sites, as the Migration trial is only available from WordPress.com.

Fixes https://github.com/Automattic/dotcom-forge/issues/2977

## Proposed changes:
1. Add the `was_migration_trial` flag to the site data
2. Add tests for `ecommerce` and `migration` trials

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* `jetpack docker phpunit -- --filter=WP_Test_Jetpack_Site_Json_Api_Endpoints --debug --verbose`
* Follow instruction found [here](https://github.com/Automattic/jetpack/pull/31907#issuecomment-1637606726)
* Install `add/was-migration-trial-flag-for-wpcom-sites` beta
* Open https://developer.wordpress.com/docs/api/console/
* Test `/sites/__ID__` and check for `was_ecommerce_trial` and `was_migration_trial` fields

zaerl committed:
[not verified] Add support for site-level was_migration_trial property

zaerl committed:
[not verified] Add changelog

zaerl committed:
[not verified] Fix: tests should return false.

zaerl committed:
[not verified] Fix: mocked function must returns false.

vishnugopal commented on composer.lock:
This really shouldn't change in this PR right?

zaerl replied:
It's automatically changed once installed. If I remember well, the change will be pruned during deploy.

vishnugopal replied:
Ack understood, thanks!

vishnugopal commented on projects/plugins/jetpack/json-endpoints/class.wpcom-json-api-get-site-endpoint.php:
```diff
@@ -579,7 +591,10 @@ protected function render_response_key( $key, &$response, $is_user_logged_in ) {
 				break;
 			case 'was_ecommerce_trial':
-				$response[ $key ] = $this->site->was_ecommerce_trial();
+				$response[ $key ] = $this->site->was_trial( self::$jetpack_enabled_trials[0] );
```
Small nit: This `[0]` `[1]` indexing looks a bit odd to me, and what if folks add the third trial in between in the list?

Can we make `$jetpack_enabled_trials` a proper dictionary with keys being constants and check for that?

zaerl replied:
Ok, sure.

vishnugopal commented on projects/plugins/jetpack/tests/php/json-api/test-class.json-api-jetpack-site-endpoints.php:
```diff
@@ -53,7 +63,44 @@ public function test_get_site() {
+		$this->assertArrayHasKey( 'jetpack_connection_active_plugins', $options );
+	}
+
+	/**
```
This comment needs revision.

vishnugopal commented on projects/plugins/jetpack/tests/php/json-api/test-class.json-api-jetpack-site-endpoints.php:
```diff
@@ -53,7 +63,44 @@ public function test_get_site() {
+
+		wp_set_current_user( $admin->ID );
+
+		$endpoint = $this->create_get_site_endpoint();
```
Is it possible to add a truthy test here as well? I'm still learning the ropes here but I believe it's possible to create a site with a migration trial now.

zaerl replied:
It checks `false` because usually in Jetpack unit tests, all the functions that are wpcom-specific return `false` and nothing more. In this case, `has_blog_sticker`.

But I can change the function and make it more smart.

vishnugopal reviewed (commented):
Some minor nits!

zaerl committed:
[not verified] Add minor changes to trails detection

zaerl committed:
[not verified] Fix: remove unused variable

zaerl committed:
Add smart has_blog_sticker function

zaerl commented:
@vishnugopal 
1. Changed `WPCOM_JSON_API_GET_Site_Endpoint::$jetpack_enabled_trials` to an associative array
2. The unit test now is more descriptive and has the `@cover` parameter
3. Changed the mock function to be more "smart" and do not simply return `false`. It uses the options system; some arbitrary data are saved during the tests. So that we can `assertTrue/False`. This will be used automatically in the future once new trials are added.

zaerl committed:
Fix: inverted array keys/values

vishnugopal commented:
@zaerl Thank you! I'm still in the middle of setting up Jetpack again, so I can't test this PR functionality. Will do once I've set it up.

zaerl committed:
Fix: undefined variable

vishnugopal commented:
Works now, approving!
![Screenshot 2023-07-18 at 3 01 34 PM](https://github.com/Automattic/jetpack/assets/533/67addbf4-6102-4d12-a605-c9eab9d1d0af)
